File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/ForbidUnsafeSymlinksRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Files;
import java.nio.file.LinkOption;
import java.nio.file.Path;
import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.Set;
import java.util.stream.Stream;

/**
 * Rule that forbids unsafe symlinks in the project root.
 */
public enum ForbidUnsafeSymlinksRule implements HarnessCheckRule {
    INSTANCE;
    private static final String CATEGORY = "forbidUnsafeSymlinks";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final JsonNode allowedNode = catNode.get("parameters").get("allowedSymlinkPairs");
        final Set<String> allowedNames = java.util.stream.StreamSupport.stream(allowedNode.spliterator(), false)
                .flatMap(pair -> java.util.stream.Stream.of(pair.get(0).asText(), pair.get(1).asText()))
                .collect(java.util.stream.Collectors.toCollection(LinkedHashSet::new));
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        final List<String> rootBases = List.of("AGENTS.md", "CLAUDE.md", "ARCHITECTURE.md", "docs", ".claude", ".github");
        return rootBases.stream()
                .flatMap(baseName -> {
                    try {
                        final Path base = root.resolve(baseName);
                        return validateSymlinks(root, base, allowedNames, severity).stream();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }

    private List<Finding> validateSymlinks(Path root, Path base, Set<String> allowedNames, String severity) throws MojoExecutionException {
        if (Files.isSymbolicLink(base)) {
            final String name = base.getFileName().toString();
            return !allowedNames.contains(name) || HarnessCheckHelper.allowedRootContractTarget(root, base) == null
                    ? List.of(new Finding(severity, CATEGORY, "symlink scan root is not allowed: " + root.relativize(base)))
                    : List.of();
        }
        return HarnessCheckHelper.safeFileOrWalk(root, base).stream()
                .filter(Files::isSymbolicLink)
                .flatMap(file -> {
                    final String name = file.getFileName().toString();
                    return !allowedNames.contains(name) || HarnessCheckHelper.allowedRootContractTarget(root, file) == null
                            ? Stream.of(new Finding(severity, CATEGORY, "symlink " + (Files.isDirectory(file, LinkOption.NOFOLLOW_LINKS) ? "directory" : "file") + " is not allowed: " + root.relativize(file)))
                            : Stream.empty();
                })
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/ForbidUnsafeSymlinksRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.isSymbolicLink
import kotlin.io.path.listDirectoryEntries
import kotlin.io.path.readSymbolicLink
import kotlin.io.path.relativeTo

/**
 * Rule that forbids disallowed symlinks at the root level.
 */
class ForbidUnsafeSymlinksRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "forbidUnsafeSymlinks"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> {
		val category = "forbidUnsafeSymlinks"
		val catObj = manifest[category]?.jsonObject ?: return emptyList()
		val parametersObj = catObj["parameters"]?.jsonObject ?: return emptyList()
		val messagesObj = catObj["messages"]?.jsonObject ?: return emptyList()
		val allowedPairs = parametersObj["allowedSymlinkPairs"]?.jsonArray?.mapNotNull { pairElem ->
			val pair = pairElem.jsonArray
			if (pair.size < 2) null else {
				val a = pair[0].jsonPrimitive.contentOrNull ?: return@mapNotNull null
				val b = pair[1].jsonPrimitive.contentOrNull ?: return@mapNotNull null
				a to b
			}
		} ?: emptyList()

		val allowed = buildSet<Pair<String, String>> {
			allowedPairs.forEach { (a, b) ->
				add(a to b)
				add(b to a)
			}
		}

		return buildSet<Finding> {
			val rootFiles = try {
				root.listDirectoryEntries()
			} catch (_: Exception) {
				emptyList()
			}
			rootFiles.filter { it.isSymbolicLink() }.forEach { file ->
				val target = try {
					file.readSymbolicLink().toString()
				} catch (_: Exception) {
					""
				}
				if (file.name to target !in allowed) {
					val msg = HarnessCheck.Companion.stringFrom(messagesObj, "fileNotAllowed").takeIf { it.isNotEmpty() } ?: "symlink file is not allowed: ${file.relativeTo(root)}"
					add(Finding(Severity.ERROR, category, msg))
				}
			}
		}.toList()
	}
}
